### PR TITLE
[JS] feat: add sso prompt

### PR DESCRIPTION
**Pull request by KennethBWSong**
#403

### CODE REVIEW

File: js/packages/teams-dialogs/src/prompts/teamsBotSsoPrompt.ts [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+       */
+      certificateContent: string;
+    }
+  );
+
+export interface TeamsBotSsoPromptSettings {
+  scopes: string[];
+  timeout?: number;
+  endOnInvalidMessage?: boolean;
+  appCredentials: TeamsBotSsoPromptAppCredentials;
```
Comment (blackchoey) [inline js/packages/teams-dialogs/src/prompts/teamsBotSsoPrompt.ts:56]: Move this out of the prompt setting and put to prompt constructor: we should ensure minimum changes to existing code

File: js/packages/teams-dialogs/src/prompts/teamsBotSsoPrompt.ts [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+import { TurnContext, TeamsChannelAccount, TeamsInfo, CardFactory, TokenExchangeResource, OAuthCard, ActionTypes, MessageFactory, Activity, ActivityTypes, verifyStateOperationName, tokenExchangeOperationName, TokenExchangeInvokeRequest, StatusCodes } from '@microsoft/teams-core';
+import { v4 as uuidv4 } from 'uuid';
+import { PromptRecognizerResult } from "./prompt";
+import { DialogContext } from "../dialogContext";
+import jwt_decode from "jwt-decode";
+import { ConfidentialClientApplication, NodeAuthOptions } from "@azure/msal-node";
+
+const invokeResponseType = "invokeResponse";
+
+type TeamsBotSsoPromptAppCredentials = {
```
Comment (blackchoey) [inline js/packages/teams-dialogs/src/prompts/teamsBotSsoPrompt.ts:16]: may rename to `TeamsBotSsoPromptAuthConfig`

**Comment by corinagum** [comment]
Hi @KennethBWSong  and @blackchoey we also have this branch: [OAuth branch](https://github.com/microsoft/teams-ai/tree/stevenic/oauthprompt) that Steve started and I'm supposed to complete. I haven't had the time to closely compare the two but I think we would want to make sure Steve's original work is matched.

**Comment by blackchoey** [comment]
@corinagum this PR is for a different story: it consolidates the TeamsBotSsoPrompt from `@microsoft/teamsfx` to the new `teams-dialogs` package. John already synced with Steve on this so @KennethBWSong creates PR for the consolidation. This won't impact the ongoing work from Steve and you.
By the way, there's also a prototype that resolves the message replay issue in [OAuth branch](https://github.com/microsoft/teams-ai/tree/stevenic/oauthprompt):https://github.com/microsoft/teams-ai/pull/369, FYI.

**Review by blackchoey — APPROVED** [review]
To do item, can be done with another PR:
1. Add logic to ensure only Teams channel can use it
2. Define types for the token response
3. Migrate existing UTs to this package

File: js/packages/teams-dialogs/src/prompts/teamsBotSsoPrompt.ts [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+    this.authConfig = authConfig;
+    this.settings = settings;
+    this.initiateLoginEndpoint = initiateLoginEndpoint;
+
+    this.validateScopesType(settings.scopes);
+  }
+
+  /**
+   * Called when a prompt dialog is pushed onto the dialog stack and is being activated.
+   * @remarks
```
Comment (corinagum) [inline js/packages/teams-dialogs/src/prompts/teamsBotSsoPrompt.ts:175]: Please change `@remarks` to `@summary` :) (remarks is deprecated and I am trying to clean up the usage of it)

**Comment by corinagum** [comment]
> @corinagum this PR is for a different story: it consolidates the TeamsBotSsoPrompt from `@microsoft/teamsfx` to the new `teams-dialogs` package. John already synced with Steve on this so @KennethBWSong creates PR for the consolidation. This won't impact the ongoing work from Steve and you. By the way, there's also a prototype that resolves the message replay issue in [OAuth branch](https://github.com/microsoft/teams-ai/tree/stevenic/oauthprompt):#369, FYI.

oh okay, sounds good! @KennethBWSong or @blackchoey, could one of you file a work item to use with this PR, and add it to the PR summary? (having a summary of changes would be great too)

**Comment by corinagum** [comment]
Could we also get unit tests before this is merged in? :) 
Working on getting this merged: https://github.com/microsoft/teams-ai/pull/402

**Comment by blackchoey** [comment]
> Could we also get unit tests before this is merged in? :) Working on getting this merged: #402

sure

**Comment by corinagum** [comment]
@blackchoey  and @KennethBWSong, more unit tests can be found at #429 -- js package should be set up well for people to add tests :)

File: js/packages/teams-dialogs/src/prompts/teamsBotSsoPrompt.ts [hunk 4]
```diff
@@ -0,0 +1,470 @@
+/**
```
Comment (corinagum) [inline js/packages/teams-dialogs/src/prompts/teamsBotSsoPrompt.ts:1]: Use same format as other tests: `____.spec.ts`?

Comment (KennethBWSong) [reply js/packages/teams-dialogs/src/prompts/teamsBotSsoPrompt.ts:1]: I am a little confused about the naming. In the pr you mentioned above, I found tests are naming like `___.spec.ts` in `teams-ai`, but under `teams-dialogs`, tests are naming like `____.test.js`. Which one should I follow?

Comment (corinagum) [reply js/packages/teams-dialogs/src/prompts/teamsBotSsoPrompt.ts:1]: Oh, interesting point... I'm assuming `teams-dialogs` are the tests directly ported from `botbuilder-dialogs`. It should be fine to keep `__.tests.js`, though I'm not sure the purpose of using js.

Thank you for pointing this out!

**Comment by corinagum** [comment]
@Stevenic could we get your review?

**Comment by corinagum** [comment]
Closing in favor of #670, which is a branch off this PR and has more updates

**Pull request closed without merging**